hs_core/hs_rdf.py

```diff
@@ -0,0 +1,358 @@
+RDFS1 = Namespace("http://www.w3.org/2000/01/rdf-schema#")
+
+NAMESPACE_MANAGER = NamespaceManager(Graph())
+NAMESPACE_MANAGER.bind('hsterms', HSTERMS, override=False)
+NAMESPACE_MANAGER.bind("rdfs1", RDFS1, override=False)
+NAMESPACE_MANAGER.bind('dc', DC, override=False)
+NAMESPACE_MANAGER.bind('dcterms', DCTERMS, override=False)
+
+
+class RDF_MetaData_Mixin(object):
+    """
+    A mixin for MetaData objects which store their metadata in generic relations.  If metadata outside of generic
+    relations need to be used, you may extend ingest_metadata and get_rdf_graph to include the other metadata elements
+    """
+
+    def rdf_subject(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_subject")
+
+    def rdf_metadata_subject(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_metadata_subject")
+
+    def rdf_type(self):
+        raise NotImplementedError("RDF_Metadata_Mixin implementations must implement rdf_type")
+
+    @classmethod
+    def rdf_subject_from_graph(cls, graph):
+        """Derive the root subject of an rdflib Graph by returning the subject in the triple with predicate DC.title"""
+        subject = None
+        for s, _, _ in graph.triples((None, DC.title, None)):
+            subject = s
+            break
+        if not subject:
+            raise Exception("Invalid rdf/xml, could not find required predicate dc:title")
+        return subject
+
+    def ingest_metadata(self, graph):
+        """Given an rdflib Graph, run ingest_rdf for all generic relations on the object"""
+        subject = self.rdf_subject_from_graph(graph)
+
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            generic_relation.related_model.ingest_rdf(graph, subject, self)
+        self.save()
+
+    def get_rdf_graph(self):
+        """adds the rdf triples of all generic relations on the object into an rdflib Graph"""
+        graph = Graph()
+        graph.namespace_manager = NAMESPACE_MANAGER
+
+        subject = self.rdf_subject()
+        # graph.add((subject, RDF.type, self.rdf_type()))
+        generic_relations = list(filter(lambda f: isinstance(f, GenericRelation), type(self)._meta.virtual_fields))
+        for generic_relation in generic_relations:
+            for f in getattr(self, getattr(generic_relation, 'name', None), None).all():
+                f.rdf_triples(subject, graph)
+
+        return graph
+
+    def get_xml(self, pretty_print=True, include_format_elements=True):
+        """Generates ORI+RDF xml for this metadata"""
+        # ORE = Namespace("http://www.openarchives.org/ore/terms/")
+        g = self.get_rdf_graph()
+        # g.add((self.rdf_metadata_subject(), ORE.describes, self.rdf_subject()))
+        return g.serialize(format='hydro-xml').decode()
```

Comment:
Commented lines can be removed. Also, it looks to me g could be an empty rdf graph, and if that is the case, will decode() call raise an exception for an empty graph?

Reply:
If the graph is empty, decode will still work as serialize will return a result, there is boilerplate in all graphs.